### ISSUE TITLE
libflux: return more clear error on uninitialized reactor in `flux_future_wait_for()`

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#4343]
This is similar to #4160.  I basically did:

```
        flux_future_t *f2 = flux_future_create (NULL, NULL);
        flux_future_set_flux (f2, h);
        if (flux_future_get (f2, NULL) < 0) {
            perror ("flux_future_get");
            exit (1);
        }
```

and you get: `flux_future_get: Invalid argument`.  The reason is in `flux_future_wait_for()` (called from `flux_future_get()`), if the timeout is < 0, a timeout watcher is never created.  So the now reactor has nothing on it, so when `flux_reactor_run()` is later executed, it just returns immediately.

Now that I've hit this twice, I'm thinking this is probably a corner case that needs to be corrected.

I see possible solutions.  An experiment putting an idle watcher into the now reactor generally speaking worked (did not test thoroughly).

```
+        if (timeout < 0.) {       // disable
            flux_watcher_stop (now->timer);
+            if (!now->idle) {
+                now->idle = flux_idle_watcher_create (now->r, NULL, NULL);
+                if (!now->idle)
+                    return -1;
+            }
+            flux_watcher_start (now->idle);
+        }
```

although I'm not 100% if having a reactor with only the idle watcher is ok?  Is the reactor basically just spinning at that point?

another option might be to set the timeout watcher to `DBL_MAX` timeout length, which would timeout in bajillions of years.

**Comment by chu11** [comment]
probably also applies to the timeout via `flux_future_then()`

**Comment by chu11** [comment]
@garlick Oh, this is just the reproducer :-)  I intend to fulfill that future, but before code can fulfill it, this is what is causing failures.

**Comment by garlick** [comment]
I think we're getting lost because `flux_future_get()` and `flux_future_wait_for()` are both normally blocking operations  if the future is unfulfilled.  If there future has to be manually fulfilled (not by some other reactor event) then since the thread is blocked, that can't happen = deadlock.

Aside: the sync flag would be processed inside the KVS not by the client right?

**Comment by chu11** [comment]
> Aside: the sync flag would be processed inside the KVS not by the client right?

Lets discuss that in #4308.

**Comment by chu11** [comment]
> The problem (I think) is that you've already called flux_future_then() on the commit RPC, so there is an active RPC on the then reactor in the handle. If the user calls a synchronous get on the placeholder future, which blocks in the now reactor, there is no way for the outstanding RPC on the then reactor to be fulfilled, so you are stuck. (This is in addition to the fact that the empty future has no watchers, so the call errors out)

Perhaps I'm missing something.  The then reactor is the `flux_t` handle's reactor and the placeholder's reactor is the temporarily created `now` reactor.  If they are two different reactors, wouldn't the above be ok?  no deadlock?  Its possible I'm missing something about how the reactor code works underneath, but I figured if they were different reactors, that they worked independently of each other.

**Comment by chu11** [comment]
Its possible I may end up solving #4308 a different way given @garlick comment above, but perhaps getting back to the question, is this a bug in `flux_future_wait_for()`?

```
        flux_future_t *f2 = flux_future_create (NULL, NULL);
        flux_future_set_flux (f2, h);
        if (flux_future_get (f2, NULL) < 0) {
            perror ("flux_future_get");
            exit (1);
        }
```

because it seems to me that this little snippet should hang, it shouldn't return an error?

OR

perhaps it should return an error, but just a more clear error that says "hey, this future isn't initialized and doesn't do anything, the wait on this future makes no sense"

**Comment by garlick** [comment]
Yeah, maybe EINVAL is too generic.

EWOULDBLOCK/EAGAIN (resource temporarily unavailable) suggests that the call should be retried, so maybe a better choice is
EDEADLOCK (resource deadlock avoided)?

**Comment by grondo** [comment]
> EDEADLOCK (resource deadlock avoided)?

Oh, good one! (Honestly I didn't know of that `errno`)

**Comment by chu11** [comment]
> That is what I was thinking. This is an error because it is a veritable deadlock. If we could detect it, then we could return a meaningful error (maybe repurpose EWOULDBLOCK?)

I thought I had an issue about this awhile back, but it'd be convenient if we could get a count of the number of watchers / thingies on the reactor.  If we can detect that the reactor "thingies" count is 0, that'd be an easy way to detect this (I think).

**Comment by chu11** [comment]
@garlick I don't think so.  The EINVAL returned is the last EINVAL listed below.

```
int flux_future_wait_for (flux_future_t *f, double timeout)
{
    if (!f) {
        errno = EINVAL;
        return -1;
    }
    if (!future_is_ready (f)) {
        if (timeout == 0.) { // don't setup 'now' context in this case
            errno = ETIMEDOUT;
            return -1;
        }
        if (!f->now) {
            if (!(f->now = now_context_create ()))
                return -1;
        }
        if (now_context_set_timeout (f->now, timeout, f) < 0)
            return -1;
        f->now->running = true;
        if (f->init && !f->now->init_called) {
            f->init (f, f->init_arg); // might set error
            f->now->init_called = true;
        }
        if (!future_is_ready (f)) {
            if (flux_reactor_run (f->now->r, 0) < 0)
                return -1; // errno set by now_timer_cb or other watcher
        }
        if (f->now->h)
            flux_dispatch_requeue (f->now->h);
        f->now->running = false;
    }
    if (!future_is_ready (f)) {
        /* Typically this error should be "impossible", but can occur
         * if the future does not get initialized properly to use the
         * now reactor.
         */
        errno = EINVAL;
        return -1;
    }
    return 0;
}
```

So I think turning this to EDEADLOCK would make sense.  Although the code is a little hard to follow, the logic isn't obvious.  I'm thinking perhaps:

```
--- a/src/common/libflux/future.c
+++ b/src/common/libflux/future.c
@@ -472,21 +472,22 @@ int flux_future_wait_for (flux_future_t *f, double timeout)
             f->now->init_called = true;
         }
         if (!future_is_ready (f)) {
-            if (flux_reactor_run (f->now->r, 0) < 0)
+            if (flux_reactor_run (f->now->r, 0) < 0) {
+                f->now->running = false;
                 return -1; // errno set by now_timer_cb or other watcher
+            }
+            if (!future_is_ready (f)) {
+                /* Typically this error should be "impossible", but can occur
+                 * if the future does not get initialized properly to use the
+                 * now reactor.
+                 */
+                errno = EDEADLOCK;
+            }
         }
         if (f->now->h)
             flux_dispatch_requeue (f->now->h);
         f->now->running = false;
     }
-    if (!future_is_ready (f)) {
-        /* Typically this error should be "impossible", but can occur
-         * if the future does not get initialized properly to use the
-         * now reactor.
-         */
-        errno = EINVAL;
-        return -1;
-    }
     return 0;
 }
```
